design.md

```diff
@@ -0,0 +1,250 @@
+
+### Scoping
+We've scoped the storage of each process to be in heap data-structures. Although this may cause some contention on the heap lock, we see it as reasonable to for the scope of this Project.
+
+# API
+
+There are two key techniques we will be using to wrap the Job Library through the API. 	We'll be using a Middleware service called `go-chi` which will allow us to do our two *tricks* with ease.
+
+1. We will be doing **``dependency enjection``** to ensure our route handlers have access to the requisite Job and Authorization Service. We do this by defining an `[Env]` structure shown below.
+
+	```go
+	type Env struct {
+		jobService Jobs.JobsManager
+		authZService sync.Map // uuid -> list
+	}
+	```
+	We will construct our handlers as follows:
+	```go 
+	 func (env *Env) createJob(w http.Response...)
+	```
+	We can subseqeuntly use the go-chi router to pass the handler with the dependent services.
+	```go
+		r.Post("/", env.createJob)
+	```
+
+	This will allow us to interact with the Jobs Manager Service provided by the env.
+2. We'll use the **CommonName** that has been verified through our CA chain and the TLS handshake to authenticate user. We will index our `authZService` using the CommonName stored in the client cert. This can be found here: `r.TLS.VerifiedChains[0][0].Subject.CommonName`.
+We will make the assumption that the CommonName will always be unique to the client and that CAs will only sign the correct clients.
+
+## Security - Authentication
+As we've shown above, we'll use the TLS mutual auth to authenticate the client and use the CommonName as the User Identity.
+
+## Security - Authorization
+Our middleware will handle Authorization.
+1. When creating a job, it will store the ID of the job into the map with the User ID as the key.
+2. When requesting access or mutation of a job, we will have middleware that checks the job ID in the param and ensures the ID matches one that the User has access to.
+
+## Router & Endpoints:
+
+### `/api/job [post]`
+This endpoint creates a new Job.
+
+### Payload requirements: 
+```graphql
+{ "cmd": !String, "args": ![String]}
+```
+
+### Response Payload - Success (200): 
+```graphql
+{ "id": uuidv4}
+```
+
+### `/api/job/:jobid [get]`
+This endpoint gets details about the job specified at jobid.
+
+This endpoint will use authorization middleware to check the user has access to the endpoint. If the user doesn't have access, it will send a `403` response and stop routing the request.
+
+### Response Payload - Success (200): 
+```graphql
+{ "cmd": !String, "args": ![String], "active": !bool, "exitCode": int}
```

Comment:
How will you represent a job that finished on its own vs was stopped by a user?

Reply:
That's a good point, I'm going to refactor the Status into some form of rich data-structure with an enum style status.